Adding bounty closure upon advanced payout.

Anish-Agnihotri opened the pull request:
##### Description

This issue aims to resolve the problem where current bounties are not closed if Advanced Payout is used.

##### Checklist
- [X] linter status: 100% pass
- [X] changes don't break existing behavior
- [X] commit message follows [commit guidelines](https://docs.gitcoin.co/mk_contributors/#step-4-commit)

##### Affected core subsystem(s)

UI + Functionality. Conferring through Slack.

##### Testing

As usual, tested across multiple platforms (MacOS + Windows), and different Wallet accounts. 

##### Refers/Fixes

N/A, issue discussed through slack.

owocki commented on app/assets/v2/js/pages/bulk_payout.js:
```diff
@@ -214,20 +214,28 @@ var update_registry = function() {
 
+  let transactions = [];
+
+  first_transaction = {
```
do we want to add something to the UI to let the funder toggle whether/not they want  to cancel the bounty?  what should that look like? @willsputra @PixelantDesign

owocki replied:
thanks for the detailed answer.. 

@PixelantDesign we had a convo yesterday about whether or not to re-introduce the checkbox..

owocki replied:
> Default to paying from escrow since this is what users expect.

I just dont know that theres any way to pay from escrow without re-adding that checkbox.  Since Tips/bounties are two seperate systems

Anish-Agnihotri replied:
Also, just for my clarification, is `pay from escrow` in reference to: 
1. The old way where users are refunded and then a new transaction made.
2. Or, where the new function is made to have payments directly subtracted from the escrow amount?

Sorry, sorry - just clarifying for my part :)

PixelantDesign replied:
@owocki I was not aware that we could not by default pay from escrow. If that is the case, we'll need a check box that says 

- [ ] Pay using bounty funds

Pay from escrow to me means automatically pulling from the bountied funds.
Defer to @owocki @mbeacom on the implementation question from @Anish-Agnihotri. 

cc: @ceresstation

owocki replied:
OK yes I think that we need to add the checkbox back @Anish-Agnihotri

Anish-Agnihotri replied:
Alright, I'll be adding that checkbox in and pushing a PR with <24hr ETA. 

I'm actually at a conference today and all of my pending PR's that have been in limbo for the past month are held up.

owocki replied:
hows it going?

PixelantDesign replied:
Thanks @Anish-Agnihotri 

My only comment is **please show the orange alert anytime checkbox is unchecked.**
It should say: Please note additional funds will be withdrawn from your Metamask wallet.


Updating the thread here for transparency: 

Anish Agnihotri [10:54 PM]
Hey, I’m working on advanced payout atm and getting it knocked out. Basically, the integration I’m going for is now:

1. If the checkbox is not checked, and total amount is under 100% payout as tip from your own wallet.

2. If the checkbox is not checked, and total amount is over 100%, payout as tip from your wallet, but also give an additional checkbox option (in the orange alert) to close bounty.

3. If the checkbox is checked, pay amount of tips from escrow amount first, and then if escrow is remaining keep task open, and if not - close the task.

Anish-Agnihotri commented:
Working on this right now. Just need to figure out a way to make payments from escrow. Everything else (UI functionality is working just perfectly). I'll keep updating this ticket as I go, and will change away from WIP status upon completion.

Anish-Agnihotri commented:
Perfect!

Thanks @PixelantDesign for clarification. To move conversation to Github: 

I already have this functionality that is needed implemented. I'll be re-adding the cancellation `first transaction` and add a tooltip hover tonight and pushing. We should finally be ready for merge.

Thanks,

Anish

Anish-Agnihotri commented:
Hey!

Just pushed out the ready-to-go commit that includes everything that is needed to have Advanced Payout functioning to spec as advised by @PixelantDesign. 

Here are the changes that have been made in detail: 

**Bulk Payout**
1. New checkbox added that checks whether or not user is looking to pay with bounty funds. Follows specs provided previously.
2. New on-hover tool-tip for 'Pay from Bounty Funds' checkbox.
3. New alert added if used unchecks 'Pay from Bounty Funds' checkbox alerting them that they will be paying as a tip, and not from the bounty amount.
4. New `first_transaction` function added that allows for closing the bounty and refunding the escrow amount.

**Tool-tip**

As you will notice, I have edited the tool-tip files to move from an ID-first system to a Class-first system. Previously, we only had one hover tool-tip which meant we could use ID's for global stylings. With a new tool-tip it's impossible to do so since one ID cannot overpower another in CSS. For this reason, I've switched all of the pre-existing ID's to Classes, and added custom ID's to style each of the tooltips.

This has been done in every location where these tool-tips are currently used. 

That's about it. Hope this is up to spec, feel free to provide feedback!

CC @owocki, @PixelantDesign

PixelantDesign commented:
@saptaks @thelostone-mc could you review please?

pinkiebell commented:
lgtm.
Two things:

Do we want to move the payout table in line with the other elements?
Personally I like it better and we avoid clipping this way.

![localhost_8000_issue_advanced_payout_pk 1856 network rinkeby](https://user-images.githubusercontent.com/40266861/48614931-bfaa3c00-e98f-11e8-8bda-ee07c72c914a.png)

clipping issue:
![localhost_8000_issue_advanced_payout_pk 1856 network rinkeby 1](https://user-images.githubusercontent.com/40266861/48614952-ce90ee80-e98f-11e8-8856-0d360733d0b5.png)


@Anish-Agnihotri 
Here is my diff, the js part should be implemented because I almost always have `web3.eth.coinbase = null` at early page load 😄 

```
diff --git a/app/assets/v2/css/bulk_payout_bounty.css b/app/assets/v2/css/bulk_payout_bounty.css
index 039be1a26..cdd47fecb 100644
--- a/app/assets/v2/css/bulk_payout_bounty.css
+++ b/app/assets/v2/css/bulk_payout_bounty.css
@@ -51,11 +51,6 @@
 
 #trans_preview {
   padding: 5px;
-  border-radius: 3px;
-  background-color: #fafafa;
-  position: absolute;
-  top: 5rem;
-  right: -20rem;
 }
 
 #trans_preview h5 {
@@ -136,4 +131,4 @@ tr div {
   color: #E66700;
   padding-bottom: 5px;
   display: none;
-}
\ No newline at end of file
+}
diff --git a/app/assets/v2/js/pages/bulk_payout.js b/app/assets/v2/js/pages/bulk_payout.js
index c3b5d0121..21cbf2861 100644
--- a/app/assets/v2/js/pages/bulk_payout.js
+++ b/app/assets/v2/js/pages/bulk_payout.js
@@ -201,7 +201,15 @@ var get_total_cost = function() {
   return total;
 };
 
-var update_registry = function() {
+var update_registry = function(coinbase) {
+  if (!coinbase) {
+    web3.eth.getCoinbase(
+      function(err, result) {
+        update_registry(result);
+      }
+    );
+    return;
+  }
 
   var num_rows = $('#payout_table tbody').find('tr').length;
   var tc = round(get_total_cost(), 2);
@@ -209,7 +217,7 @@ var update_registry = function() {
   var original_amount = $('#original_amount').val();
   var net = round(original_amount - tc, 2);
   var over = round((original_amount - get_total_cost()) * -1, 4);
-  var addr = web3.eth.coinbase.substring(38);
+  var addr = coinbase.substring(38);
   var pay_with_bounty = $('#pay_with_bounty').is(':checked');
   
   $('#total_cost').html(tc + ' ' + denomination);
```

PixelantDesign commented:
Hey Pinkiebell! Thanks for the review. 

I think we could definitely figure out a better placement for display. The immediate need is getting this in and preventing the payout issues we've been having. 

I'd love to get this in and tested and think about the preview in a separate ticket.

Thanks for the great work @Anish-Agnihotri

Anish-Agnihotri commented:
Thanks for the review and feedback @pinkiebell! That makes sense for sure, and I'll have a chat with @PixelantDesign about including that in (seems we'll be doing it in a new ticket). 

Think we're ready to merge, then 🥂.

owocki commented:
@Anish-Agnihotri my QA: 

new flyout is :100:

http://bits.owocki.com/438c39b0f6f0/Screen%20Shot%202018-11-16%20at%209.51.40%20AM.png

javascript error on pageload http://bits.owocki.com/e60507bfc4ef/Screen%20Shot%202018-11-16%20at%209.52.28%20AM.png

also i can’t click the ‘add (+)’ button.. it seems that its not wired up

this is the bounty https://stage.gitcoin.co/issue/advanced_payout?pk=5199&network=rinkeby

pinkiebell commented:
@owocki 
You have the same error , Hah.
@Anish-Agnihotri 
Can you please add the js part from my patch I posted earlier ? That fixes the error .

PixelantDesign commented:
Looks pretty good to me aside from the things that @owocki and @pinkiebell mentioned above.

Thanks @Anish-Agnihotri!

Anish-Agnihotri commented:
Ah, that does seem like an issue. Thank you for solution @pinkiebell! Will push the change as soon as I get home in ~2hrs.

owocki commented:
@Anish-Agnihotri let us know!  antsy to push this

Anish-Agnihotri commented:
@owocki @pinkiebell 

Just pushed out the newest changes according to @pinkiebell, tyvm. I can't actually test the changes on my end, since my environment won't let me send a transaction for some reason. 

I'd appreciate if someone could check it out, and just give it the thumbs up!

- Anish

pinkiebell commented:
LGTM 👍

owocki commented:
just pushed a fix d02b80308e579d65f3e3f76b79404d1ee954c71b

Anish-Agnihotri commented:
Much appreciated @owocki. 

Apologies for being unable to knock that out myself this weekend.

Glad to see it finally in place :)

Anish

owocki commented:
Thanks again for the help @Anish-Agnihotri - i dont see a bounty on this work at all but lmk if i missed it

Anish-Agnihotri commented:
@owocki 

Much appreciated, but certainly not necessary in anyway. This was bountied a while back in #2294, although that issue was closed, and these were just improvements on the same base. Due to that, I'd be glad to refund this tip, since most of the things that were done within this PR were simply to build upon the last one.

Thanks again,

Anish

owocki commented:
@Anish-Agnihotri you contribute in a lot of places (not just this PR), so please keep the tip!
